src/DotVVM.Framework/Compilation/DotvvmViewCompilationService.cs

```diff
@@ -0,0 +1,202 @@
+                    {
+                        if (routes == null)
+                            routes = dotvvmConfiguration.RouteTable.Select(r => new DotHtmlFileInfo() {
+                                VirtualPath = r.VirtualPath,
+                                Url = r.Url,
+                                HasParameters = r.ParameterNames.Any(),
+                                DefaultValues = r.DefaultValues.Select(s => s.Key + ":" + s.Value?.ToString()).ToList(),
+                                RouteName = r.RouteName,
+                                Status = (string.IsNullOrWhiteSpace(r.VirtualPath) || !IsDotvvmPresenter(r))
+                                    ? CompilationState.NonCompilable
+                                    : CompilationState.None
+                            }).ToList();
+                    }
+                }
+            }
+            return routes;
+        }
+
+        public DotvvmViewCompilationService(DotvvmConfiguration dotvvmConfiguration)
+        {
+            this.dotvvmConfiguration = dotvvmConfiguration;
+        }
+
+        private static object loadMasterPagesLocker = new object();
+        private void InitMasterPagesCollection()
+        {
+            if (masterPages == null)
+            {
+                lock (loadMasterPagesLocker)
+                {
+                    if (masterPages == null) masterPages = new List<DotHtmlFileInfo>();
+                }
+            }
+        }
+        
+        private bool IsDotvvmPresenter(RouteBase r)
+        {
+            var presenter = r.GetPresenter(dotvvmConfiguration.ServiceProvider);
+            return presenter.GetType().IsInstanceOfType(dotvvmConfiguration.RouteTable.GetDefaultPresenter(dotvvmConfiguration.ServiceProvider));
+        }
+
+        public async Task<bool> CompileAll(bool buildInParallel=true, bool forceRecompile = false)
+        {
+            List<DotHtmlFileInfo> controlsToCompile;
+            List<DotHtmlFileInfo> routesToCompile;
+            if (forceRecompile)
+            {
+                routesToCompile = routes;
+                controlsToCompile = controls;
+            }
+            else
+            {
+                routesToCompile = routes.Where(t => t.Status == CompilationState.None).ToList();
+                controlsToCompile = controls.Where(t => t.Status == CompilationState.None).ToList();
+            }
+            
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+            var compileTasks = routesToCompile.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList();
+
+            compileTasks.AddRange(controlsToCompile.Select(a => new Task(() => BuildView(a, tempMasterPages))).ToList());
+
+            await ExecuteCompileTasks(compileTasks,buildInParallel);
+
+            while (tempMasterPages.Count > 0)
+            {
+                masterPages = masterPages.Union(tempMasterPages).Distinct(DotHtmlFileInfo.VirtualPathComparer).ToList();
```

Comment:
Are you aware that this might be running in parallel with another `CompileAll`? Please also see the review of https://github.com/riganti/dotvvm-samples-compilation-status-page/pull/12, but we definitely don't want to assign mutable fields just like this.